### PR TITLE
feat: introduce retry logic with exponential backoff in batch upload flow

### PR DESCRIPTION
**Pull request by SKannaniOS**
## Description
This PR introduces sophisticated retry logic with exponential backoff strategy for the batch upload flow in RudderStackAnalytics Swift SDK. The implementation adds resilience against transient network failures while maintaining proper error handling for non-retryable errors.

The retry mechanism uses an exponential backoff with jitter to prevent thundering herd problems and includes a cool-off period after maximum retry attempts to prevent excessive API calls. The system intelligently differentiates between retryable errors (network issues, server errors) and non-retryable errors (client errors like 400, 401, 404, 413) to ensure appropriate handling of different failure scenarios.

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactor/optimization

## Implementation Details
- **Backoff Policy System**: Implemented a protocol-based backoff policy system with `BackoffPolicy` protocol and `ExponentialBackoffPolicy` implementation
- **Retry Loop**: Added retry logic in `EventUploader.uploadBatch()` that continues retrying for `RetryableEventUploadError` types
- **Exponential Backoff with Jitter**: Delay calculation follows `delay = minDelay * base^attempt + randomJitter` formula to prevent synchronized retries
- **Cool-off Period**: After 5 failed attempts, enters a 30-minute cool-off period before resuming retries
- **Thread-Safe State Management**: Uses Swift's Actor pattern (`BackoffPolicyHandler`) for safe concurrent access to retry state
- **Error Classification**: Maintains existing error handling for non-retryable errors (400, 401, 404, 413) while adding retry logic for retryable errors
- **Project Structure**: Reorganized `Policies` folder with `FlushPolicy` and `BackoffPolicy` subfolders for better organization
- **Comprehensive Logging**: Added verbose logging for retry attempts, backoff delays, and cool-off periods for monitoring and debugging

### Configuration:
- **Base multiplier**: 2.0
- **Minimum delay**: 3 seconds
- **Maximum attempts**: 5 attempts before cool-off
- **Cool-off period**: 30 minutes (1800 seconds)
- **Jitter**: Random jitter applied to prevent thundering herd

## Checklist
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

### Automated Testing:
- **Unit Tests**: Run `BackoffPolicyTests` suite which covers:
  - Exponential backoff policy behavior and reset functionality
  - Backoff handler state management and delay application
  - Mock-based testing with proper test doubles
  - Edge cases including multiple resets and boundary conditions

### Manual Testing:
1. **Network Failure Simulation**: 
   - Disconnect network during event upload to verify retry behavior
   - Observe exponential backoff delays in logs
   - Verify successful upload when network is restored

2. **Server Error Simulation**:
   - Mock 5xx server errors to test retry logic
   - Verify backoff is applied between retry attempts
   - Check cool-off period behavior after max attempts

3. **Non-Retryable Error Testing**:
   - Test 400, 401, 404, 413 error responses
   - Verify no retry attempts are made for these errors
   - Confirm appropriate error handling (batch deletion, analytics shutdown, etc.)

## Breaking Changes
None. This is a backward-compatible enhancement that doesn't affect the public API. All existing functionality remains unchanged.

## Maintainers Checklist
- [ ] The code has been reviewed.
- [ ] CI tests have passed.
- [ ] All necessary documentation has been updated.

## Screenshots (if applicable)
N/A - This is an internal enhancement without UI changes.

## Additional Context
- **Architecture Decision**: Used Actor pattern for thread-safe state management instead of locks or serial queues
- **Performance Impact**: Minimal memory overhead; delays are expected during network issues (intentional behavior)
- **Future Extensibility**: Protocol-based design allows easy addition of different backoff strategies (linear, fixed, custom)

### Files Added:
- `Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift`
- `Sources/RudderStackAnalytics/Policies/BackoffPolicy/ExponentialBackoffPolicy.swift`
- `Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicyHandler.swift`
- `Tests/RudderStackAnalyticsTests/BackoffPolicyTests.swift`

### Files Modified:
- `Sources/RudderStackAnalytics/EventQueue/EventUploader.swift` - Added retry loop and backoff integration
- `RudderStackAnalytics.xcodeproj/project.pbxproj` - Updated project structure and build configuration

### Files Moved:
- Moved flush policy files to `Sources/RudderStackAnalytics/Policies/FlushPolicy/` subdirectory for better organization

### Sample Values (in milliseconds):
5750, 6005, 21994, 35631, 67114...

### CODE REVIEW

File: Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+import Foundation
+
+// MARK: - BackoffPolicy
+/**
+ Protocol representing a backoff policy for retrying operations.
+ Implementations should provide a strategy for calculating the next delay
+ and resetting the backoff state.
+ */
+protocol BackoffPolicy {
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift:16]:
```suggestion
// MARK: - BackOffPolicy
/**
 Protocol representing a backoff policy for retrying operations.
 Implementations should provide a strategy for calculating the next delay
 and resetting the backoff state.
 */
protocol BackOffPolicy {
```

File: Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    /**
+     Calculates the next delay in seconds based on the backoff policy.
+     
+     - Returns: The next delay in seconds.
+     */
+    func nextDelayInSeconds() -> Int
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift]: It should be in `millis` to align with the Kotlin interface, refer [BackOffPolicy - Kotlin interface](https://github.com/rudderlabs/rudder-sdk-kotlin/blob/3c521e175b88cd3f20b20362a734a1d010d4bfc5/core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/policies/backoff/BackOffPolicy.kt#L15).
```suggestion
    /**
     Calculates the next delay in milliseconds based on the backoff policy.
     
     - Returns: The next delay in milliseconds.
     */
    func nextDelayInMillis() -> Int
```

CC: @vgupta98

File: Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ A struct containing constants used in backoff policies.
+ */
+struct BackOffPolicyConstants {
+
+    private init() {
+        /* Default implementation (no-op) */
+    }
+
+    static let maxAttempts = 5
+    static let coolOffPeriodInSeconds = 1800 // 30 minutes
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift]: Nice to add a comment. It helps in readability.

File: Sources/RudderStackAnalytics/Policies/BackoffPolicy/ExponentialBackoffPolicy.swift [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    /**
+     Calculates the next delay in seconds based on the exponential backoff policy.
+     
+     - Returns: The next delay in seconds with jitter applied.
+     */
+    func nextDelayInSeconds() -> Int {
+        let delayInSeconds = Int(Double(minDelayInSecs) * pow(base, Double(attempt)))
+        attempt += 1
+        
+        let delayWithJitterInSeconds = withJitter(delayInSeconds)
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/Policies/BackoffPolicy/ExponentialBackoffPolicy.swift]: Delay should be in `milliseconds` and not `seconds`.

Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/Policies/BackoffPolicy/ExponentialBackoffPolicy.swift]: Refer to the Kotlin implementation for more details.

Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/Policies/BackoffPolicy/ExponentialBackoffPolicy.swift]: Please update the whole class to use `milliSeconds` references (e.g., change the method name from `nextDelayInSeconds` to `nextDelayInMillis` or `nextDelayInMilliSeconds`.

File: Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicyHandler.swift [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+import Foundation
+
+// MARK: - BackoffPolicyHandler
+/**
+ Actor is responsible for managing the backoff policy for retrying operations.
+ */
+final actor BackoffPolicyHandler {
+    private let maxAttempts: Int
+    private let coolOffSecs: Int
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicyHandler.swift]: Shall we consider renaming this variable to:
```suggestion
    private let coolOffPeriodInSecs: Int
```

**Review by 1abhishekpandey — CHANGES_REQUESTED** [review]
Nice. Let's refactor the BackOff policy to use `MilliSeconds` instead of just `seconds` and also update at other relevant places to use the millis values.

File: Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift (hunk 2 above)
Comment (vgupta98) [reply Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift]: We are keeping the delay as well the jitter which is added to it in milliseconds.

File: Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift (hunk 1 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift:16]: `Backoff` is treated as a single word in most programming contexts (similar to “retry” or “timeout”). `BackOffPolicy` with the capital “O” looks like two separate words and is less idiomatic in Swift. So I used `BackoffPolicy`.

Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift:16]: Let's rename [this](https://github.com/rudderlabs/rudder-sdk-swift/blob/b58e7d29dff42b36eaa77db8db2b85f3c4225c86/Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift#L36) then:
> struct BackOffPolicyConstants {

Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/Policies/BackoffPolicy/BackoffPolicy.swift:16]: It's already renamed.

**Review by 1abhishekpandey — APPROVED** [review]

**Review by vgupta98 — APPROVED** [review]
